### PR TITLE
Allow to use custom keys in RedisRegistrationStore

### PR DESCRIPTION
**Pull request by aliakseiz**
My attempt to implement requirements mentioned in https://github.com/eclipse/leshan/issues/1249 

Not sure about the prefixes, though. Left them untouched for now, but suggestions are still welcome.

### CODE REVIEW

**Review by sbernard31 — COMMENTED** [review]
I will try to propose some javadoc for builder setter in another review

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java [hunk 1]
```diff
@@ -768,4 +754,273 @@ public void run() {
...
+            this.endpointExpirationKey = endpointExpirationKey;
+            return this;
+        }
+
+        /**
+         * Set the clean period in seconds.
+         * <p>
+         * Default value is {@literal 60 seconds}.
+         */
+        public Builder setCleanPeriod(long cleanPeriod) {
```
Comment (sbernard31) [inline leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java:890]: Maybe something like :

Set time between 2 periodic task about cleaning expired registration.

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java [hunk 2]
```diff
@@ -768,4 +754,273 @@ public void run() {
...
+            this.cleanPeriod = cleanPeriod;
+            return this;
+        }
+
+        /**
+         * Set the maximum number to clean in a clean period.
+         * <p>
+         * Default value is {@literal 500}.
+         */
+        public Builder setCleanLimit(int cleanLimit) {
```
Comment (sbernard31) [inline leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java:900]: Maybe something like :

Set maximum number of expired registration removed by clean period

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java [hunk 3]
```diff
@@ -768,4 +754,273 @@ public void run() {
...
+            this.cleanLimit = cleanLimit;
+            return this;
+        }
+
+        /**
+         * Set the grace period in seconds.
+         * <p>
+         * Default value is {@literal 0 seconds}.
+         */
+        public Builder setGracePeriod(long gracePeriod) {
```
Comment (sbernard31) [inline leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java:910]: Maybe something like  : 

Set some extra time to add to registration lifetime when calculating if a registration expired.

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java [hunk 4]
```diff
@@ -768,4 +754,273 @@ public void run() {
...
+            this.gracePeriod = gracePeriod;
+            return this;
+        }
+
+        public Builder setSchedExecutor(ScheduledExecutorService schedExecutor) {
```
Comment (sbernard31) [inline leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java:915]: Maybe something like :

Set ScheduledExecutorService used to launch period task about cleaning expired registration.

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java [hunk 5]
```diff
@@ -768,4 +754,273 @@ public void run() {
...
+            this.schedExecutor = schedExecutor;
+            return this;
+        }
+
+        public Builder setLock(JedisLock lock) {
```
Comment (sbernard31) [inline leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java:920]: For this one, I think it's maybe ok to not have javadoc or maybe something like : 

Set JedisLock implementation used to handle concurrent access to this store.
Default implementation used is `{@link SingleInstanceJedisLock}`

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java [hunk 6]
```diff
@@ -768,4 +754,273 @@ public void run() {
...
+            this.lock = lock;
+            return this;
+        }
+
+        public Builder setRegistrationSerDes(RegistrationSerDes registrationSerDes) {
```
Comment (sbernard31) [inline leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java:925]: Maybe something like :

Set `{@link RegistrationSerDes}` instance used to serialize/de-serialize registration to/from this store.

**Review by sbernard31 — COMMENTED** [review]
My separated review about javadoc

**Comment by sbernard31** [comment]
> Not sure about the prefixes, though. Left them untouched for now, but suggestions are still welcome.

That sounds good to me :+1: 

Maybe out of topic but : 
Reading this PR, and especially this builder method : 
`public Builder setRegistrationSerDes(RegistrationSerDes registrationSerDes)`

I was thinking that maybe we should also have a setter to serialize/de-serialize `Observation` ?
`public Builder setObservationSerDes(ObservationSerDes observationSerDes)`

**Comment by sbernard31** [comment]
Could you also rebase your PR on `master` because I fixed a github action issue : #1410.  :pray:

**Comment by sbernard31** [comment]
@aliakseiz let me know when you plan to finalize this because I would like to declare a code freeze on `master` soon because of M11 release.
Potentially, If you can not spend more time on it, I can finished work on this PR. (most is already done)

**Comment by sbernard31** [comment]
I will start to work on this.

**Comment by sbernard31** [comment]
@aliakseiz I created a new PR fixing comment above. See #1414.

I plan to integrate it into master before the end of the day. 
If you have time, you could review it. 
(If you can not, it's not a be deal, sorry about the rush this is pretty unusual :disappointed:)

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java [hunk 7]
```diff
@@ -768,4 +754,273 @@ public void run() {
...
+
+    /**
+     * Class helping to build and configure a {@link RedisRegistrationStore}.
+     */
+    public static class Builder {
+        /** Default time in seconds between 2 cleaning tasks (used to remove expired registration) */
+        public static final long DEFAULT_CLEAN_PERIOD = 60;
+        public static final int DEFAULT_CLEAN_LIMIT = 500;
+        /** Default extra time for registration lifetime in seconds */
+        public static final long DEFAULT_GRACE_PERIOD = 0;
```
[the remark this answers is not in the document]
Comment (aliakseiz) [reply leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java:766]: Totally makes sense. Just a copy-paste issue.

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java [hunk 8]
```diff
@@ -104,35 +99,25 @@ public class RedisRegistrationStore implements RegistrationStore, Startable, Sto
...
-    public RedisRegistrationStore(Pool<Jedis> p, ScheduledExecutorService schedExecutor, long cleanPeriodInSec,
-            long lifetimeGracePeriodInSec, int cleanLimit) {
-        this(p, schedExecutor, cleanPeriodInSec, lifetimeGracePeriodInSec, cleanLimit, new SingleInstanceJedisLock());
-    }
-
-    public RedisRegistrationStore(Pool<Jedis> p, ScheduledExecutorService schedExecutor, long cleanPeriodInSec,
-            long lifetimeGracePeriodInSec, int cleanLimit, JedisLock redisLock) {
-        this(p, schedExecutor, cleanPeriodInSec, lifetimeGracePeriodInSec, cleanLimit, redisLock,
-                new RegistrationSerDes());
+        this(new Builder(p));
```
[the remark this answers is not in the document]
Comment (aliakseiz) [reply leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java:102]: Right, somehow I missed that.
`generateDefaultValue()` sounds good to me. I will implement it the same way.

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java (hunk 1 above)
Comment (aliakseiz) [reply leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java:890]: I am not very familiar with the Leshan internals yet, so it's difficult for me to write descriptive JavaDocs. It was mostly a copy-paste of already existing comments.
Thank you for suggestions, helped me to better understand it.

**Comment by sbernard31** [comment]
This PR is part of #1414 and is now integrated in `master`.

@aliakseiz thx a lot for your contributions :pray: !

**Comment by sbernard31** [comment]
@aliakseiz could I ask which part and how you are currently using Leshan ? 

Would you like to add your company as [Leshan adopters](https://www.eclipse.org/leshan/#who%20is%20using) ? (see https://github.com/eclipse/leshan/issues/830)

**Comment by aliakseiz** [comment]
@sbernard31 we have a LwM2M gateway heavily based on the Leshan, which imports `core` and `server` packages.
Major differences relate to SecurityServlet, observation logic and, till recently, to Redis storage. Several thousands of devices both secure and insecure are connected to it. Sending measurements primarily over UDP.

We also have multiple variations of LwM2M clients, that are based on Leshan and Wakaama.

> Would you like to add your company as [Leshan adopters](https://www.eclipse.org/leshan/#who%20is%20using) ? (see https://github.com/eclipse/leshan/issues/830)

That would be very nice actually. I will leave a comment there.

**Comment by sbernard31** [comment]
@aliakseiz thx for details explanation. 

> That would be very nice actually. I will leave a comment there.

That would be great. :+1:  
Once you add the comment  at https://github.com/eclipse/leshan/issues/830 and if you agree I will create a PR at [eclipsefdn-project-adopters](https://gitlab.eclipse.org/eclipsefdn/it/api/eclipsefdn-project-adopters) to add your company at Leshan adopters.

**Comment by aliakseiz** [comment]
Actually I already did: https://gitlab.eclipse.org/eclipsefdn/it/api/eclipsefdn-project-adopters/-/issues/193

Does it still make sense to leave a comment in #830?

**Comment by sbernard31** [comment]
> Actually I already did: https://gitlab.eclipse.org/eclipsefdn/it/api/eclipsefdn-project-adopters/-/issues/193
 
Oops I missed that :grin: 

> Does it still make sense to leave a comment in https://github.com/eclipse/leshan/issues/830?

It makes sense but clearly not mandatory.

Some users add both.
Some only add to Leshan adopters. 
Some are OK to put comment in PR but doesn't want to be in Leshan adopters.

All is fine too me.

**Comment by sbernard31** [comment]
(I created a PR for it : https://gitlab.eclipse.org/eclipsefdn/it/api/eclipsefdn-project-adopters/-/merge_requests/203)

**Pull request closed without merging**